### PR TITLE
Feature/maroon 319 coord system

### LINKED ISSUE #319: [Coord System] Setup
Adding a coordinate system object which can be used for experiments like coulombs law.

### PR DESCRIPTION
**Pull request by dpdsw**
closes #319

### CODE REVIEW

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/SystemChangeHandler.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using UnityEngine;
+
+namespace Maroon.Physics.CoordinateSystem
+{
+/*
+ * Currently doing nothing
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/SystemChangeHandler.cs]: Do we still need this script?
If not, please remove it.

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystem.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            var systemSpaceCoordinates = WorldCoordinatesToSystemSpace(objectTransform);
+
+            var xValue = _axisDictionary[Axis.X].GetValueFromAxisPoint(systemSpaceCoordinates.x);
+            var yValue = _axisDictionary[Axis.Y].GetValueFromAxisPoint(systemSpaceCoordinates.y);
+            var zValue = _axisDictionary[Axis.Z].GetValueFromAxisPoint(systemSpaceCoordinates.z);
+
+            return new Vector3(xValue, yValue, zValue);
+        }
+
+        public Vector3 GetPositionInWorldSpace(Vector3 localSpacePosition, Unit[] respectiveUnits)
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystem.cs]: Can we make the respectiveUnity parameter optional?
e.g. GetPositionInWorldSpace(Vector3 localSpacePosition, Unit[] respectiveUnits =null)
respectiveUnits ??= new[] { Unit.m, Unit.m, Unit.m };

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+            }
+            else
+            {
+                _spaceIndicator.gameObject.transform.localScale = _enableThirdDimension
+                    ? new Vector3(axisWorldLengths.x, axisWorldLengths.y, axisWorldLengths.z)
+                    : new Vector3(axisWorldLengths.x, axisWorldLengths.y, 0.2f);
+            }
+        }
+
+        public Vector3 GetWorldLengthsOfDirection(bool positive)
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs]: Can we use C# proporties for this?
e.g. 
Vector3 PositiveWorldLength => new Vector3(.....);
Vector3 NegativeWorldLength => new Vector3(...);

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+                { Axis.X, _axisList.ElementAt(FindInList(Axis.X)) },
+                { Axis.Y, _axisList.ElementAt(FindInList(Axis.Y)) },
+                { Axis.Z, _axisList.ElementAt(FindInList(Axis.Z)) },
+                { Axis.NX, _axisList.ElementAt(FindInList(Axis.NX)) },
+                { Axis.NY, _axisList.ElementAt(FindInList(Axis.NY)) },
+                { Axis.NZ, _axisList.ElementAt(FindInList(Axis.NZ)) }
+            };
+        }
+
+        private int FindInList(Axis direction)
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs]: Why do we need this function? Just use linq functions.
e.g.  _axisList.Find(axis => axis.AxisID == Axis.X);

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                _uniformWorldAxisLength = value;
+                SetAxisWorldLengthUniform();
+            }
+        }
+
+        #endregion
+
+        private void Awake()
+        {
+            _axisDictionary = DictionaryFromAxisList();
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs]: Why not just:
foreach (var axis in _axisList)
  _axisDictionary[axis.AxisID] = axis;

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Editor/CoordAxisEditor.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+        public override void OnInspectorGUI()
+        {
+            EditorGUILayout.LabelField("General", EditorStyles.boldLabel);
+
+            EditorGUILayout.PropertyField(_id);
+            EditorGUILayout.PropertyField(_prefab);
+
+            EditorGUI.BeginChangeCheck();
+            EditorGUILayout.PropertyField(_worldLength);
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Editor/CoordAxisEditor.cs]: Can we use a slider for this property?

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Editor/CoordAxisEditor.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            serializedObject.ApplyModifiedProperties();
+            if (EditorGUI.EndChangeCheck())
+            {
+                axis.SetupAxis();
+            }
+
+            EditorGUILayout.LabelField("Subdivision of Axis", EditorStyles.boldLabel);
+
+            EditorGUI.BeginChangeCheck();
+            EditorGUILayout.PropertyField(_axisSubdivision);
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Editor/CoordAxisEditor.cs]: Can we use a slider for this property? And allow only integer numbers?

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Editor/CoordAxisEditor.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            serializedObject.ApplyModifiedProperties();
+            if (EditorGUI.EndChangeCheck())
+            {
+                axis.SetupAxis();
+            }
+
+            EditorGUILayout.LabelField("Local Axis Length/Unit", EditorStyles.boldLabel);
+
+            EditorGUI.BeginChangeCheck();
+            EditorGUILayout.PropertyField(_localLength);
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Editor/CoordAxisEditor.cs]: Can we use a slider for this property?

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Editor/CoordAxisEditor.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        public override void OnInspectorGUI()
+        {
+            EditorGUILayout.LabelField("General", EditorStyles.boldLabel);
+
+            EditorGUILayout.PropertyField(_id);
+            EditorGUILayout.PropertyField(_prefab);
+
+            EditorGUI.BeginChangeCheck();
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Editor/CoordAxisEditor.cs]: one BeginChangeCheck at the start and one EndChangeCheck at the end should be enough.

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordAxis.cs [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+        private void Start()
+        {
+            SetupAxis();
+        }
+
+        public void UpdateFontSize(float value)
+        {
+            _fontSize = value;
+            foreach (var marker in GetComponentsInChildren<TextMeshPro>())
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordAxis.cs]: why not in _axisMarkers?

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordAxis.cs [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+                    _direction = Vector3.left;
+                    break;
+                case Axis.NY:
+                    _direction = Vector3.down;
+                    break;
+                case Axis.NZ:
+                    _direction = Vector3.back;
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordAxis.cs]: Debug.LogWarning() or Debug.LogError() should be enough?

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+        nm = -9,
+        µm = -6,
+        mm = -3,
+        cm = -2,
+        dm = -1,
+        m = 0,
+        km = 3
+    }
+
+    public class CoordSystemManager : MonoBehaviour
```
Comment (michaelholly) [inline unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs]: Can we rename this class to AxisManager/Controller or something similar

**Review by michaelholly — CHANGES_REQUESTED** [review]

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/SystemChangeHandler.cs (hunk 1 above)
Comment (dpdsw) [reply unity/Assets/Maroon/reusablePrefabs/CoordSystem/SystemChangeHandler.cs]: File + dependencies removed

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystem.cs (hunk 2 above)
Comment (dpdsw) [reply unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystem.cs]: "respectiveUnits" is now optional, calculation is adjusted accordingly ( null & Unit.None are now calculated as m)

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs (hunk 3 above)
Comment (dpdsw) [reply unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs]: Yes, done.

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs (hunk 4 above)
Comment (dpdsw) [reply unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs]: Complicated thinking

File: unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs (hunk 5 above)
Comment (dpdsw) [reply unity/Assets/Maroon/reusablePrefabs/CoordSystem/Scripts/Runtime/CoordSystemManager.cs]: Complicated thinking

**Review by michaelholly — APPROVED** [review]

**Pull request merged**